src/engine/datafacade/mmap_memory_allocator.cpp

```diff
@@ -16,47 +17,52 @@ namespace engine
 namespace datafacade
 {
 
-MMapMemoryAllocator::MMapMemoryAllocator(const storage::StorageConfig &config,
-                                         const boost::filesystem::path &memory_file)
+MMapMemoryAllocator::MMapMemoryAllocator(const storage::StorageConfig &config)
 {
     storage::Storage storage(config);
+    std::vector<storage::SharedDataIndex::AllocatedRegion> allocated_regions;
 
-    if (!boost::filesystem::exists(memory_file))
     {
-        storage::DataLayout initial_layout;
-        storage.PopulateStaticLayout(initial_layout);
-        storage.PopulateUpdatableLayout(initial_layout);
-
-        auto data_size = initial_layout.GetSizeOfLayout();
-
-        storage::io::BufferWriter writer;
-        storage::serialization::write(writer, initial_layout);
-        auto encoded_layout = writer.GetBuffer();
-
-        auto total_size = data_size + encoded_layout.size();
-
-        mapped_memory = util::mmapFile<char>(memory_file, mapped_memory_file, total_size);
-
-        std::copy(encoded_layout.begin(), encoded_layout.end(), mapped_memory.data());
-
-        index = storage::SharedDataIndex(
-            {{mapped_memory.data() + encoded_layout.size(), std::move(initial_layout)}});
-
-        storage.PopulateStaticData(index);
-        storage.PopulateUpdatableData(index);
+        std::unique_ptr<storage::BaseDataLayout> fake_layout =
+            std::make_unique<storage::TarDataLayout>();
+
+        // Convert the boost::filesystem::path object into a plain string
+        // that's stored as a member of this allocator object
+        rtree_filename = storage.PopulateLayoutWithRTree(fake_layout);
+
+        // Now, we add one more AllocatedRegion, with it's start address as the start
+        // of the rtree_filename string we've saved.  In the fake_layout, we've
+        // stated that the data is at offset 0, which is where the string starts
+        // at it's own memory address.
+        // The syntax &(rtree_filename[0]) gets the memory address of the first char.
+        // We can't use the convenient `.data()` or `.c_str()` methods, because
+        // prior to C++17 (which we're not using), those return a `const char *`,
+        // which isn't compatible with the `char *` that AllocatedRegion expects
+        // for it's memory_ptr
+        allocated_regions.push_back({&(rtree_filename[0]), std::move(fake_layout)});
```

Comment:
Why does AllocatedRegion expect a `char *` then in the first place?

Reply:
Historical reasons I believe - I've fixed and changed the stored base pointers to `void *`